File: tests/conformance/probes/notifications.py
#!/usr/bin/env python3
"""Notifications' one job: be the only place a notification on this machine can land, and still
have it after a restart.

There were four notification systems on this desktop and none knew about the others.

`mako` held `org.freedesktop.Notifications`, so `notify-send`, Chromium and every ordinary Linux
program drew popups in mako's style that nothing of ours could see. The shell ALSO implemented
that interface, from a thread of its own, and raced mako for the name — whichever won got the
machine's notifications and the other half went dark; the audit of 17 September caught mako
winning by a few hundred milliseconds. The shell kept its own private JSON file in `~/.yantrik/`,
fed by screenshots and focus mode, that no service and no mind could read. And this service —
autostarted, with five methods — had an in-memory `Vec` that **nothing in the whole tree ever
wrote to**, so it was always empty and nobody noticed.

So the checks below are all one question asked five ways: did it land in the ONE store, and is it
still there. Both doors are exercised, because the point is that they are the same door:
`notify-send` (freedesktop, the way every foreign program arrives) and `yos notify` (ours). The
restart is the check the old service could not have passed at all.

Leave-as-found is by DISMISSING what this probe posted, not by restoring the file. The store is
live and shared: a notification that arrived from somewhere else while this ran is real, and
putting an old file back over it would destroy it. Dismissed notifications stay in the file for a
week by design, so "removed" here means "no longer showing", which is what dismiss means
everywhere else on this machine.
"""

import json
import os
import pathlib
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import lib  # noqa: E402

ONE_JOB = ("Take a notification from any sender on this machine — ours or any ordinary Linux "
           "program — into one store, count it where the desktop can see it, keep it across a "
           "restart of the service, and let it be dismissed.")

APP = "notifications"
SERVICE_BIN = "/opt/yantrik/bin/notifications-service"
SERVICE_SOCK = lib.SOCKET_DIR / "notifications.sock"
STORE = pathlib.Path.home() / ".local/share/yantrik/notifications.json"

# Unique per run, so a probe that crashed last time cannot be mistaken for this one's work and
# so two probes on one machine cannot dismiss each other's notifications.
MARK = "conformance-%d-%d" % (os.getpid(), int(time.time()))
FREEDESKTOP_TITLE = "Probe via notify-send %s" % MARK
YANTRIK_TITLE = "Probe via yos notify %s" % MARK


def listed():
    """Everything the store says is showing, newest first."""
    try:
        return lib.call(APP, "notifications.list") or []
    except SystemExit:
        return []


def mine(items=None):
    """Only the notifications this run posted."""
    items = listed() if items is None else items
    return [n for n in items if MARK in (n.get("title") or "")]


def find(title, items=None):
    for n in (listed() if items is None else items):
        if n.get("title") == title:
            return n
    return None


def dismiss_mine():
    """Take back everything this probe put there. Safe to call twice."""
    removed = []
    for n in mine():
        try:
            lib.call(APP, "notifications.dismiss", {"id": n["id"]})
            removed.append(n["id"])
        except SystemExit:
            pass
    return removed


def service_up():
    return lib.surface_up(APP)


def start_service():
    """Ask the desktop to start it, the way anything else on this machine would."""
    asked = lib.act("shell", "start_service", name=APP)
    lib.wait_until(service_up, timeout=10, what="the notifications service to answer")
    return asked


def run():
    """The probe. In a function so that importing this file does nothing — see calendar.py."""
    not_exercised = []

    with lib.Probe(APP, ONE_JOB) as probe:
        probe.note("processes_before", lib.running(SERVICE_BIN))
        probe.note("store_path", str(STORE))
        probe.note("marker", MARK)

        try:
            # ── 1. The one store is answering ─────────────────────────────
            if not service_up():
                start_service()
            up = lib.wait_until(service_up, timeout=15, what="the notifications service")
            probe.check(
                "the notifications service answers on its socket",
                bool(up),
                contract=1,
                evidence=up.evidence(socket=str(SERVICE_SOCK),
                                     processes=lib.running(SERVICE_BIN)))
            if not up:
                # Nothing below can be checked, and saying so once beats nine identical
                # failures that all mean "it is not running".
                probe.note("stopped_early",
                           "The service never answered, so no sender could be tested.")
                return

            state = lib.state(APP)
            probe.note("describe_before", state)
            freedesktop = str(state.get("freedesktop", ""))
            holds_the_bus = freedesktop.startswith("serving")
            probe.check(
                "the service holds org.freedesktop.Notifications, or names who has it instead",
                holds_the_bus or "owned by" in freedesktop or "unavailable" in freedesktop,
                contract=4,
                evidence={"freedesktop": freedesktop})

            # ── 2. An ordinary Linux program ──────────────────────────────
            #
            # notify-send is libnotify's own client and knows nothing about Yantrik. If it lands
            # here, so does Chromium, and so does any script anybody writes.
            if not holds_the_bus:
                not_exercised.append(
                    "a notify-send lands in the store with source=freedesktop")
                probe.note("freedesktop_not_exercised", {
                    "statement": "The freedesktop door was NOT EXERCISED: this service does not "
                                 "hold org.freedesktop.Notifications, so notify-send reached "
                                 "whoever does. On a machine that has not taken the update that "
                                 "is mako, still started from the labwc autostart.",
                    "freedesktop": freedesktop,
                })
            else:
                sent = lib.run_and_wait(
                    ["notify-send", "--app-name=ConformanceProbe", "--urgency=low",
                     FREEDESKTOP_TITLE, "posted by the conformance probe"],
                    timeout=20)
                probe.note("notify_send", sent)
                landed = lib.wait_until(lambda: find(FREEDESKTOP_TITLE), timeout=10,
                                        what="the notify-send notification to reach the store")
                from_fd = landed.value if landed else None
                probe.check(
                    "a notify-send lands in the store with source=freedesktop",
                    bool(from_fd) and from_fd.get("source") == "freedesktop",
                    contract=8,
                    evidence={"stored": from_fd, "app": (from_fd or {}).get("app"),
                              "source": (from_fd or {}).get("source"),
                              "waited": landed.evidence() if landed else None})
                probe.check(
                    "the sender's own name is kept, not replaced by ours",
                    bool(from_fd) and from_fd.get("app") == "ConformanceProbe",
                    contract=3,
                    evidence={"app": (from_fd or {}).get("app")})

            # ── 3. Our own one-liner ──────────────────────────────────────
            sent = lib.run_and_wait(
                [lib.YOS, "notify", YANTRIK_TITLE, "posted by the conformance probe",
                 "--app", "ConformanceProbe"],
                timeout=20)
            probe.note("yos_notify", sent)
            landed = lib.wait_until(lambda: find(YANTRIK_TITLE), timeout=10,
                                    what="the yos notify notification to reach the store")
            from_us = landed.value if landed else None
            probe.check(
                "`yos notify` lands in the same store, with source=yantrik",
                bool(from_us) and from_us.get("source") == "yantrik",
                contract=8,
                evidence={"stored": from_us, "source": (from_us or {}).get("source"),
                          "waited": landed.evidence() if landed else None})
            if not from_us:
                probe.note("stopped_early",
                           "Nothing was stored, so the restart and dismiss checks below would "
                           "be about an empty store rather than about keeping anything.")
                return

            # ── 4. The desktop can see it ─────────────────────────────────
            #
            # The whole reason there is one store: the shell's own account of the machine has to
            # agree with it. The shell polls once a second, so this is given a moment.
            def shell_unread():
                block = lib.describe("shell").get("state", {}).get("notifications") or {}
                return block.get("unread")

            counted = lib.wait_until(lambda: (shell_unread() or 0) > 0, timeout=8,
                                     what="the desktop to count the unread notification")
            shell_block = lib.describe("shell").get("state", {}).get("notifications") or {}
            probe.note("shell_notifications", shell_block)
            probe.check(
                "the desktop's own describe reports the unread count from the same store",
                bool(counted) and isinstance(shell_block.get("unread"), int)
                and shell_block["unread"] > 0,
                contract=8,
                evidence={"shell_notifications": shell_block,
                          "waited": counted.evidence() if counted else None})
            probe.check(
                "the desktop names the most recent notifications, not just a number",
                any(MARK in (n.get("title") or "") for n in (shell_block.get("latest") or [])),
                contract=3,
                evidence={"latest": shell_block.get("latest")})

            # ── 5. Kill it and start it again ─────────────────────────────
            #
            # The check the old service could not have passed: its store was a `Vec` in memory,
            # so everything on this machine died with the process — and the shell autostarts it,
            # which meant every reboot.
            before = {n["id"]: n["title"] for n in mine()}
            probe.note("mine_before_restart", before)
            killed = lib.kill_app(SERVICE_BIN)
            try:
                SERVICE_SOCK.unlink(missing_ok=True)
            except OSError:
                pass
            probe.note("killed", killed)
            time.sleep(1)
            start_service()
            back = lib.wait_until(service_up, timeout=15,
                                  what="the notifications service to come back")
            probe.check(
                "the service comes back when it is asked for",
                bool(back),
                contract=1,
                evidence=back.evidence(processes=lib.running(SERVICE_BIN)))

            after = {n["id"]: n["title"] for n in mine()} if back else {}
            probe.note("mine_after_restart", after)
            probe.check(
                "what was stored is still there after the service is restarted",
                bool(before) and before == after,
                contract=6,
                evidence={"before": before, "after": after,
                          "store_exists": STORE.exists(),
                          "store_bytes": STORE.stat().st_size if STORE.exists() else None})

            # The ids are the same ones, which is what lets a sender close its own notification
            # across a restart — the freedesktop spec promises exactly that.
            probe.check(
                "the ids survive too, so a sender can still close what it sent",
                bool(before) and set(before) == set(after),
                contract=6,
                evidence={"before_ids": sorted(before), "after_ids": sorted(after)})

            # ── 6. Dismiss ────────────────────────────────────────────────
            target = next(iter(after or before), None)
            if target:
                answer = lib.act(APP, "dismiss", id=target)
                probe.note("dismiss", answer)
                # Checked against the store's own list, not against the action's answer: an
                # action that reports what it did not do is the fault this whole suite exists
                # for, and this service's `dismiss` used to answer success over an id it had
                # never seen.
                still_listed = [n["id"] for n in listed() if n["id"] == target]
                probe.check(
                    "dismissing one notification takes it off the list",
                    not still_listed,
                    contract=2,
                    evidence={"dismissed": target, "answer": answer.get("result"),
                              "refused": answer.get("refused"),
                              "still_listed": still_listed})

            # A refusal names its reason rather than reporting success over nothing.
            missing = lib.act(APP, "dismiss", id="this-id-does-not-exist")
            probe.check(
                "dismissing something that is not there is refused, in words that say why",
                bool(missing.get("refused")) and "this-id-does-not-exist" in missing["refused"],
                contract=3,
                evidence={"refused": missing.get("refused"),
                          "kind": lib.refusal_kind(missing)})

        finally:
            # ── Put the machine back ──────────────────────────────────────
            removed = dismiss_mine()
            probe.note("dismissed_on_the_way_out", removed)

        leftover = mine()
        probe.check(
            "nothing this probe posted is still showing",
            not leftover,
            contract="leave-as-found",
            evidence={"leftover": leftover})

        running_after = lib.running(SERVICE_BIN)
        probe.note("processes_after", running_after)
        probe.check(
            "the notifications service is running afterwards",
            bool(running_after),
            contract="leave-as-found",
            evidence={"before": probe.notes["processes_before"], "after": running_after})

        if not_exercised:
            probe.note("checks_not_exercised", not_exercised)


if __name__ == "__main__":
    run()

File: crates/yantrik-ipc-contracts/src/notifications.rs
//! Notifications contract — the one store every notification on this machine lands in.
//!
//! ## What this used to be
//!
//! A `Notification` struct with `icon`, `source_app` and a `timestamp` string, plus a
//! `NotificationService` trait that nothing on this machine implemented and nothing called. The
//! struct was used by exactly one file — the service's own `main.rs` — and the trait by none, so
//! the "contract" named one side of a conversation that had no other side.
//!
//! It is a contract now because there are several senders: the shell, our own apps, the calendar
//! service's reminder thread, `yos notify`, and every ordinary Linux program through
//! `org.freedesktop.Notifications`. They all build the same request and the store parses it once.
//!
//! ## Why every field is bounded
//!
//! Anything on the machine can post here, including programs we did not write and did not test.
//! A 10 MB body from a misbehaving web page would be written to disk, read back at every start
//! and drawn into a toast. So the store truncates on the way in — see [`clamp`] and the `MAX_*`
//! constants — and the bound is part of the contract rather than a defensive check the service
//! might forget, because the freedesktop path cannot refuse: the spec has no error reply for
//! "too long", and refusing would make our desktop the one where `notify-send` mysteriously
//! fails.

use serde::{Deserialize, Serialize};

// ── Methods ─────────────────────────────────────────────────────────────────────────────────

/// Post a notification. Params: [`AddRequest`]. Answers the stored [`Notification`].
pub const ADD: &str = "notifications.add";
/// Everything the store holds that is not dismissed, newest first. Answers `Vec<Notification>`.
pub const LIST: &str = "notifications.list";
/// What changed since a revision. Params: `{ revision: u64 }`. Answers [`Since`].
pub const SINCE: &str = "notifications.since";
/// Dismiss one. Params: `{ id }`.
pub const DISMISS: &str = "notifications.dismiss";
/// Dismiss everything currently showing. Params: `{}`.
pub const DISMISS_ALL: &str = "notifications.dismiss_all";
/// Mark one — or, with no `id`, everything — as read. Params: `{ id? }`.
pub const MARK_READ: &str = "notifications.mark_read";
/// Invoke an action a notification published. Params: `{ id, action_id }`.
pub const ACTION: &str = "notifications.action";

// ── Bounds ──────────────────────────────────────────────────────────────────────────────────

/// Longest title kept. A title is one line in a toast; past this it is elided on screen anyway.
pub const MAX_TITLE: usize = 200;
/// Longest body kept. Two or three paragraphs — enough for a real message, not a log dump.
pub const MAX_BODY: usize = 2_000;
/// Longest sender name kept.
pub const MAX_APP: usize = 64;
/// Most actions kept on one notification. The toast draws them in a row; four is the row.
pub const MAX_ACTIONS: usize = 4;
/// Longest action id kept.
pub const MAX_ACTION_ID: usize = 64;
/// Longest action label kept. A button, not a sentence.
pub const MAX_ACTION_LABEL: usize = 48;
/// How many notifications the store keeps. Oldest are dropped first.
pub const MAX_STORED: usize = 500;
/// How long a dismissed notification is kept before it is pruned. A week, so "what did that
/// installer say on Tuesday" is still answerable and the file does not grow forever.
pub const DISMISSED_TTL_SECS: i64 = 7 * 24 * 60 * 60;

/// Cut a string to `max` characters — characters, not bytes, so this cannot split a UTF-8
/// sequence and write a file that will not parse. Appends nothing: an elision marker in stored
/// data is a lie about what the sender said, and the UI elides for display on its own.
pub fn clamp(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        return s.to_string();
    }
    s.chars().take(max).collect()
}

// ── Types ───────────────────────────────────────────────────────────────────────────────────

/// How loud a notification is. The word is lowercase on the wire because that is what
/// `notify-send --urgency=` uses, what `yos notify` takes, and what the shell's describe prints.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum Urgency {
    Low,
    #[default]
    Normal,
    Critical,
}

impl Urgency {
    /// The short word, for a view or a log line.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Low => "low",
            Self::Normal => "normal",
            Self::Critical => "critical",
        }
    }

    /// Parse the word, anything unrecognised being `normal`.
    ///
    /// Deliberately total. This reads user input from `yos notify --urgency`, from an app's
    /// one-liner and from a JSON-RPC caller; the alternative to a default is refusing to deliver
    /// a notification because somebody typed "high", which loses the message to protect a field.
    pub fn parse(s: &str) -> Self {
        match s.trim().to_ascii_lowercase().as_str() {
            "low" | "0" => Self::Low,
            "critical" | "urgent" | "2" => Self::Critical,
            _ => Self::Normal,
        }
    }

    /// The freedesktop `urgency` hint byte (spec 1.2 §1.3): 0 low, 1 normal, 2 critical.
    pub fn from_hint_byte(b: u8) -> Self {
        match b {
            0 => Self::Low,
            2 => Self::Critical,
            _ => Self::Normal,
        }
    }

    /// The same byte going out, for `GetCapabilities` consumers and for the shell's toast code,
    /// which has drawn urgency as 0/1/2 since before any of this existed.
    pub fn hint_byte(self) -> u8 {
        match self {
            Self::Low => 0,
            Self::Normal => 1,
            Self::Critical => 2,
        }
    }
}

/// Where a notification came in from. Not the same thing as `app`: `app` is who is speaking,
/// this is which door they came through, and it is the difference between "Chromium said this
/// over D-Bus" and "our download manager called the service".
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum Source {
    /// Posted through `notifications.add` by something of ours.
    #[default]
    Yantrik,
    /// Arrived over `org.freedesktop.Notifications`.
    Freedesktop,
}

impl Source {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Yantrik => "yantrik",
            Self::Freedesktop => "freedesktop",
        }
    }
}

/// A button on a notification.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NotificationAction {
    /// What the sender will be told was pressed. For a freedesktop notification this is the
    /// action key that goes back out in `ActionInvoked`. For one of ours it is the name of an
    /// action on the sending app's own control surface.
    pub id: String,
    /// What the button says.
    pub label: String,
    /// Arguments for that control-surface action.
    ///
    /// The freedesktop spec has `ActionInvoked` to tell a sender its button was pressed. Our own
    /// apps have no such signal — an app that posted a notification and then exited has nothing
    /// listening at all — so for a `yantrik` notification the shell presses the button *on the
    /// sender's behalf*, through the control surface the app already publishes, starting the app
    /// if it is not running. That needs the arguments, and this is where they ride: Download
    /// Manager's "Open folder" is `open_folder` with `{"id": 4}`, which is the same call a
    /// person makes by clicking the button in its window.
    ///
    /// Without this, a button on one of our notifications would be a control that does nothing,
    /// which this repo forbids.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub args: Option<serde_json::Value>,
}

impl NotificationAction {
    /// A button that takes no arguments.
    pub fn new(id: impl Into<String>, label: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            label: label.into(),
            args: None,
        }
    }

    /// A button that calls the sending app's action with these arguments.
    pub fn with_args(mut self, args: serde_json::Value) -> Self {
        self.args = Some(args);
        self
    }
}

/// One notification, as stored and as handed to anyone who asks.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Notification {
    /// The store's own id, a decimal counter as a string: "1", "2", …
    ///
    /// A number rather than a uuid because the freedesktop spec's `Notify` has to answer with a
    /// `u32` the caller can later pass to `CloseNotification` or as `replaces_id`. With uuids
    /// the service had to keep a second id space and a map between them, and any gap between the
    /// two is a notification that cannot be closed by the program that sent it.
    pub id: String,
    /// Who is speaking, as the person would recognise them: "Downloads", "Chromium", "Calendar".
    pub app: String,
    pub title: String,
    pub body: String,
    pub urgency: Urgency,
    /// RFC 3339, UTC.
    pub created_at: String,
    pub read: bool,
    /// Dismissed notifications stay in the file for [`DISMISSED_TTL_SECS`] so that `since` can
    /// tell a client they went away, and so history survives closing a toast.
    pub dismissed: bool,
    pub actions: Vec<NotificationAction>,
    pub source: Source,
    /// The id this one replaced, when the sender asked to update rather than add.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub replaces_id: Option<String>,
    /// Who this machine established sent it — beside `app`, which is only what was said.
    ///
    /// Optional on read so a store written before this existed still loads, and absent on a
    /// freedesktop notification, whose door has not asked the bus who was behind it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sender: Option<Sender>,
    /// The store revision at which this notification last changed. `since` compares against it.
    pub revision: u64,
}

/// Who sent a notification, as this machine established it — kept beside `app`, which is only
/// what the sender said.
///
/// Found on 22 September 2026 (#114): a mind posted `app: "Yantrik"` with a body that was false
/// in every particular, and the stored record had nothing in it but what that caller had
/// written. The approval card had already been through this (#43) and answers with two lines it
/// refuses to merge — the name the caller gave itself, and the program the kernel says opened
/// the socket. A notification now carries the same two, and the shell draws them in the same
/// words.
///
/// None of this is set by the request. The service fills it in from `SO_PEERCRED` at the
/// moment the call arrives, which is the only moment the peer is certainly still there.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct Sender {
    /// `app` exactly as the caller gave it. `None` when it gave none and `app` was filled in
    /// from the verified program — so a reader can tell "it called itself Downloads" from "it
    /// was download-manager".
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub claimed: Option<String>,
    /// The program that opened the socket, as the shell's approval card prints it: the first
    /// thing in its ancestry a person would recognise, with its pid — or `could not be
    /// identified`, never a blank and never a guess.
    pub verified: String,
    /// The pid that line is about. `0` when nothing was established.
    #[serde(default)]
    pub pid: i32,
    /// Its `/proc/<pid>/exe`, empty when unknown. The line shows the command line, which says
    /// more; this is the path a person checks afterwards.
    #[serde(default)]
    pub exe: String,
}

/// What a sender posts.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AddRequest {
    pub app: String,
    pub title: String,
    #[serde(default)]
    pub body: String,
    #[serde(default)]
    pub urgency: Urgency,
    #[serde(default)]
    pub actions: Vec<NotificationAction>,
    #[serde(default)]
    pub source: Source,
    /// Replace this notification instead of adding another. Used by progress-style senders and
    /// by the freedesktop `replaces_id` argument.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub replaces_id: Option<String>,
}

/// The answer to [`SINCE`]: where the store is now, and everything that changed to get there.
///
/// `changed` includes notifications that were dismissed or marked read, not only new ones —
/// a client that only heard about additions would keep drawing a toast the person just closed
/// from the notification centre.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Since {
    pub revision: u64,
    pub changed: Vec<Notification>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn urgency_words_round_trip() {
        for u in [Urgency::Low, Urgency::Normal, Urgency::Critical] {
            assert_eq!(Urgency::parse(u.as_str()), u);
            assert_eq!(Urgency::from_hint_byte(u.hint_byte()), u);
        }
    }

    #[test]
    fn unknown_urgency_is_normal_not_an_error() {
        // A notification with a typo in one field still has to be delivered.
        assert_eq!(Urgency::parse("high"), Urgency::Normal);
        assert_eq!(Urgency::parse(""), Urgency::Normal);
        assert_eq!(Urgency::from_hint_byte(200), Urgency::Normal);
    }

    #[test]
    fn urgency_is_lowercase_on_the_wire() {
        assert_eq!(
            serde_json::to_string(&Urgency::Critical).unwrap(),
            "\"critical\""
        );
        assert_eq!(
            serde_json::from_str::<Urgency>("\"low\"").unwrap(),
            Urgency::Low
        );
    }

    #[test]
    fn a_record_written_before_the_sender_existed_still_loads() {
        // `~/.local/share/yantrik/notifications.json` on every machine that took the update
        // holds records with exactly these fields and nothing about who sent them. A field that
        // failed to parse would empty the whole store at the next start.
        let old = r#"{"id":"134","app":"Yantrik","title":"Studio finished","body":"","urgency":"normal","created_at":"2026-09-23T00:43:53Z","read":true,"dismissed":true,"actions":[],"source":"yantrik","revision":237}"#;
        let n: Notification = serde_json::from_str(old).expect("an old record parses");
        assert_eq!(n.sender, None);
        // And it is written back without inventing one.
        assert!(!serde_json::to_string(&n).unwrap().contains("sender"));

        let mut n = n;
        n.sender = Some(Sender {
            claimed: Some("Yantrik".into()),
            verified: "python -m hermes_cli.main gateway run (pid 689)".into(),
            pid: 689,
            exe: "/home/yantrik/.hermes/hermes-agent/venv/bin/python".into(),
        });
        let text = serde_json::to_string(&n).unwrap();
        let back: Notification = serde_json::from_str(&text).unwrap();
        assert_eq!(back.sender, n.sender);
    }

    #[test]
    fn clamp_counts_characters_not_bytes() {
        // Four characters, twelve bytes. Cutting at a byte offset here would write invalid
        // UTF-8 into the store and the next start would fail to parse the whole file.
        let devanagari = "नमस्ते";
        let cut = clamp(devanagari, 3);
        assert_eq!(cut.chars().count(), 3);
        assert!(devanagari.starts_with(&cut));
        assert_eq!(clamp("short", 100), "short");
    }
}
